File: Stack/394.decode-string.cpp
/*
 * @lc app=leetcode id=394 lang=cpp
 * Method 1： Stack
 * [Notice] loop times may bigger than 10 32[a2[c]]
 * [394] Decode String
 * Time Limit Exceeded O(n²)
 * 
*/
class Solution {
public:
    string decodeString(string s) {
        stack<string> str; //use string stack
        stack<int> num;
        int count=0;
        string loop = "";

        for(int i=0; i<s.size();i++){
            if(s[i]>='0' && s[i]<='9'){
                count = count*10 + s[i]-'0';
            }
            else if(s[i]=='['){
                num.push(count);
                cout<<"meent '[', push count= "<< count
                    <<" loop"=loop<<endl;
                count = 0;
                str.push(loop);
                loop.clear();
            }
            else if(s[i]==']'){
                int times = num.top();
                num.pop();
                while(times-1){
                    str.top()+=loop;
                }
                cout<<"now top
                loop = str.top();
                str.pop();
            }
            else { loop+=s[i];}
            
        }

        return str.empty() ? loop : str.top();
    }
};

File: Stack/394.decode-string.java
/*
 * @lc app=leetcode id=394 lang=java
 * Note: Stack:
    * peek(): top without return
    * pop(): top and return 
 * [394] Decode String
 */

// @lc code=start
class Solution {
    public  String decodeString(String s) {
        int n = s.length();
        StringBuilder curr = new StringBuilder();
        Stack<Integer> number = new Stack<>();
        Stack<String> word = new Stack<>();

        for(int i=0;i<n;i++){
            char c = s.charAt(i);
            if(Character.isDigit(c)){
                int num = c-'0';
                while(i<n && Character.isDigit(s.charAt(i+1))){
                    num *=10;
                    num +=s.charAt(i+1)-'0';
                    i++;
                }
                number.push(num);
            }
            else if(c=='['){ //renew curr, curr record the last []
                word.push(curr.toString()); //push curr [null,a]
                curr = new StringBuilder();
            }
            else if(c==']'){ //update curr
                StringBuilder tmp = new StringBuilder(word.pop()); //a
                int time = number.pop(); //
                while(time>0){
                    tmp.append(curr);
                    time--;
                }
                curr = tmp; //curr record the last []
            }
            else{ //letter
                curr.append(s.charAt(i));
            }
        }
        return curr.toString();
    }
}
// @lc code=end

